[Mobile Payments] Move onboarding checks to PreflightController

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

As part of the refactor of the card payments classes for Tap to Pay, we planned to move the responsibility for the onboarding check to the new `CardPresentPaymentsPreflightController`. This PR makes that change.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Using a US-based WCPay or Stripe store:

1. Navigate to `Menu > Settings > Experimental features`
2. Turn on `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. Observe that if you tap `Card` quickly, you see onboarding briefly ("Connecting to your account"), but if you wait a few seconds you don't see that before the payment.

### On your site's wp-admin plugin page, deactivate WCPay and/or Stripe

1. Navigate to `Menu > Settings > Experimental features`
2. Turn on `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. Observe that you see the `Activate WooCommerce Payments` onboarding screen 
6. Reactivate the plugin on your site
7. Tap `Refresh after updating` in the app
8. Observe that you can complete the payment.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Move onboarding checks to Preflight

joshheald commented:
@toupper No rush on this, I've marked it for 12.2

toupper commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentPreflightController.swift:
```diff
@@ -35,10 +35,19 @@ final class CardPresentPaymentPreflightController {
     ///
     private let analytics: Analytics
 
+    /// Root View Controller
```
A super not a blocker 😅  Do we need to repeat the property name in the comment? e.g `/// Root View Controller`

joshheald replied:
Not really. It's just for consistency with the rest, which is inherited from the code that's been copied and refactored from the existing flows. It's weird if we have the comments for some of the properties but not others. I don't think we have a very consistent style through the app for this though.

toupper commented:
Hi @joshheald, the code looks good, but when testing I found a few issues:

> Observe that if you tap Card quickly, you see onboarding briefly ("Connecting to your account"), but if you wait a few seconds you don't see that before the payment

. ❌  I didn't see the onboarding.

Then I disabled the WCPay plugin -without restarting the app- in my store and proceed to repeat the payment with TTP iPhone, which was successful. When processing the payment I got an error:

<img src="https://user-images.githubusercontent.com/1864060/215114953-09819b49-4501-478e-9770-2dd276cf8a05.PNG" width="375">

I then repeated the process by restarting the app, but when selected Card in the Payment Methods screen, nothing happened in screen. I got this error in the Xcode console:

`WooCommerce[4153:1683898] ⛔️ Payment Gateway not found, can't collect payment.`

Let me know if you need more information.

joshheald commented:
Thanks for the review @toupper, it would definitely be helpful to check a few things...

> > Observe that if you tap Card quickly, you see onboarding briefly ("Connecting to your account"), but if you wait a few seconds you don't see that before the payment
> 
> . ❌ I didn't see the onboarding.

Perhaps you didn't tap `Card` soon enough? (Or just have a faster connection than me?!) 

We aim not to actually show this "Connecting to your account" screen, because we do the onboarding check in the background when the Payment Method screen loads. You not seeing it isn't necessarily an issue, it may just be passing before you tap the `Card` row.

> Then I disabled the WCPay plugin -without restarting the app- in my store and proceed to repeat the payment with TTP iPhone, which was successful. When processing the payment I got an error:

When you say repeat the payment... did you leave the app in the same payment flow here? Or start again from the Orders List or Collect Payment stage?

If it was the same payment flow, then you'd be piggybacking on the previous onboarding pass, and then failing server-side, which explains the error you saw. I think that would be existing behaviour, and not a situation we'd try to specially handle.

> I then repeated the process by restarting the app, but when selected Card in the Payment Methods screen, nothing happened in screen. I got this error in the Xcode console:
> 
> `WooCommerce[4153:1683898] ⛔️ Payment Gateway not found, can't collect payment.`

This was a surprising but super helpful thing that you've found. It'll take some more work to resolve.

That error's thrown when we set up the payment, and this test has (I think) identified that we're relying on a side-effect of onboarding to correctly set up the CardPresentPaymentsStore. I'm digging in to how we can improve this now.

toupper commented:
Thanks for your answers @joshheald 

> When you say repeat the payment... did you leave the app in the same payment flow here? Or start again from the Orders List or Collect Payment stage?
> 
> If it was the same payment flow, then you'd be piggybacking on the previous onboarding pass, and then failing server-side, which explains the error you saw. I think that would be existing behaviour, and not a situation we'd try to specially handle

Yeah, it was the same payment flow. If I exit the Payments Menu and enter again, I see the _Activate WooCommerce Payments_ error screen:

<img src="https://user-images.githubusercontent.com/1864060/215513502-06f12c17-9888-43d9-b518-a8d3b185d179.PNG" width="375">

> This was a surprising but super helpful thing that you've found. It'll take some more work to resolve.
> 
> That error's thrown when we set up the payment, and this test has (I think) identified that we're relying on a side-effect of onboarding to correctly set up the CardPresentPaymentsStore. I'm digging in to how we can improve this now.

Great, let me know if you need more help here.

joshheald committed:
Remove unneccesary ProximityReader import

joshheald committed:
Rely on Onboarding to set the Payment Gateway

One part of the onboarding flow is to set the PaymentGateway on the store. This involves asking the user, if required. Setting it elsewhere can lead us to rely on side effects, and is not required

joshheald committed:
Prefer () -> Void over () -> ()

joshheald committed:
Use PaymentGateway from onboarding during payment

Previously we fetched and passed around the Payment Gateway independently of running onboarding. This meant that when we moved onboarding to be handled in Preflight, the PaymentGatewayAccount would not always have been set, and any existing value could be left over from a previous payment.

This removes paymentGatewayAccount as a passed in property, and relies on it being set in onboarding, before being stored as state (in the use case) much like the connected reader is.

A follow up to this change would be to remove this stored state and pass the values to the various tracking functions, however doing that in this branch made the change too large.

joshheald committed:
Combine CollectOrderPaymentUseCase errors

joshheald commented:
@toupper That's fixed now. I'm planning to further improve it by removing the property for `paymentGatewayAccount` in the `CollectOrderPaymentUseCase`, and pass it where required instead. That would have made this PR too big, so I'll do it next/soon.

The main issue was that we relied on `CardPresentPaymentAction.use` being dispatched in several places, and the selected gateway being passed around too much, so we had side effects which were exposed by the refactor.

Now we rely on Onboarding to set the payment gateway, and pass it around a lot less, so it should not be stale.

[Edit] You'll still see the PaymentCaptureError if you disable the WCPay plugin and retry payment without restarting the app, though. This is expected and correct, but now your next attempt at taking a payment will show the onboarding failure correctly.

joshheald committed:
Make error available for legacy flow again

toupper commented:
@joshheald Thanks for your fix! Unfortunately it compiles with errors, all stating the same, cannot find `CollectOrderPaymentUseCaseError` in `LegacyCollectOrderPaymentUseCase`. E.g.:


```
woocommerce-ios/WooCommerce/Classes/ViewRelated/Orders/Collect Payments/LegacyCollectOrderPaymentUseCase.swift:280:39: cannot find 'CollectOrderPaymentUseCaseError' in scope
--
  |  
  | onCompletion(.failure(CollectOrderPaymentUseCaseError.flowCanceledByUser))
```

joshheald commented:
@toupper Oops sorry: I'd pushed before I committed the last fix 🤦 .
Should work now.

toupper reviewed (approved):
LGTM! 🎉 🚢 :shipit:
